### PR TITLE
Addition of Normal Distribution to ANN module

### PR DESCRIPTION
**Pull request by nishantkr18**
Hey!
The implementation of Normal Distribution in https://github.com/mlpack/mlpack/pull/1912 had a small bug in the `LogProbablity()` function. Also, the backward pass and tests for the class was not present. 
This is an attempt to complete the work, as normal distribution is of use in many places, especially for continuous action space environments in RL module.
For the tests, I have taken 4 random values of `mean`, `sigma` and `x` to produce corresponding LogProbabilites and gradients, and compared it with pytorch's implementation.
I use this snippet to get the `prob`, `dmu` and `dsigma`.
```python
mu = torch.tensor(1.1, requires_grad = True)
std = torch.tensor(0.1, requires_grad = True)
n = Normal(mu, std)
action_tensor = torch.tensor(1.05, requires_grad = False)
log_prob = n.log_prob(action_tensor)
ratio = torch.exp(log_prob)

print(log_prob.data.numpy())
ratio.backward()
print('gradients of mu and sigma: ', mu.grad.data.numpy(), std.grad.data.numpy())

```
Please have a look and share your thoughts :)

### CODE REVIEW

File: src/mlpack/tests/ann_dist_test.cpp [hunk 1]
```diff
@@ -127,4 +128,40 @@ BOOST_AUTO_TEST_CASE(JacobianBernoulliDistributionLogisticTest)
...
+
+  ann::NormalDistribution normalDist =
+    ann::NormalDistribution(mu, sigma);
+
+  arma::vec x = {1.05, 1.1, 1.7, 2.5};
+
+  arma::vec prob;
+  normalDist.LogProbability(x, prob);
+
+  // testing output of log probablity for some random mu, sigma and x.
```
Comment (zoq) [inline src/mlpack/tests/ann_dist_test.cpp:147]:
```suggestion
  // Testing output of log probability for some random mu, sigma and x.
```

File: src/mlpack/tests/ann_dist_test.cpp [hunk 2]
```diff
@@ -127,4 +128,40 @@ BOOST_AUTO_TEST_CASE(JacobianBernoulliDistributionLogisticTest)
...
+  BOOST_REQUIRE_CLOSE(prob[0], 1.2586464, 1e-3);
+  BOOST_REQUIRE_CLOSE(prob[1], 0.8751131, 1e-3);
+  BOOST_REQUIRE_CLOSE(prob[2], -0.30579138, 1e-3);
+  BOOST_REQUIRE_CLOSE(prob[3], -5.498411, 1e-3);
+
+  arma::vec dmu, dsigma;
+  normalDist.ProbBackward(x, dmu, dsigma);
+
+  // testing output of dmu and dsigma for some random mu, sigma and x.
```
Comment (zoq) [inline src/mlpack/tests/ann_dist_test.cpp:156]:
```suggestion
  // Testing output of dmu and dsigma for some random mu, sigma and x.
```

File: src/mlpack/methods/ann/dists/normal_distribution.hpp [hunk 3]
```diff
@@ -0,0 +1,151 @@
+/**
+ * @file normal_distribution.hpp
+ * @author xiaohong ji
```
Comment (zoq) [inline src/mlpack/methods/ann/dists/normal_distribution.hpp:3]: Feel free to add you as well.

File: src/mlpack/methods/ann/dists/normal_distribution.hpp [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+/**
+ * Implementation of the Normal Distribution function.
+ *
+ * Normal distribution is a function which accepts a mean and a standard deviation
+ * term and creates a probablity distribution out of it.
+ */
+class NormalDistribution
+{
+ private:
```
Comment (zoq) [inline src/mlpack/methods/ann/dists/normal_distribution.hpp:29]: Can we put this at the bottom, just for consistency with the rest of the codebase.

File: src/mlpack/methods/ann/dists/normal_distribution.hpp [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+  //! Mean of the distribution.
+  arma::vec mean;
+  //! Standard deviation of the distribution.
+  arma::vec sigma;
+
+  // pi
+  static const constexpr double pi = 3.14159265358979323846264338327950288;
```
Comment (zoq) [inline src/mlpack/methods/ann/dists/normal_distribution.hpp:36]: You can use `M_PI` which is defined in `src/mlpack/prereqs.hpp`.

File: src/mlpack/methods/ann/dists/normal_distribution.hpp [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+  /**
+   * Default constructor, which creates a Normal distribution with zero
+   * dimension.
+   */
+  NormalDistribution();
+
+  /**
+   * Create a Normal distribution with the given mean and sigma.
+   */
+  NormalDistribution(const arma::vec& mean, const arma::vec& sigma);
```
Comment (zoq) [inline src/mlpack/methods/ann/dists/normal_distribution.hpp:48]: Would be nice to comment on each parameter using `@param `.

File: src/mlpack/methods/ann/dists/normal_distribution.hpp [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+  /**
+   * Stores the gradient of the probabilities of the observations
+   * with respect to mean and standard deviation.
+   *
+   * @param observation The observation matrix.
+   * @param dmu The gradient with respect to mean.
+   * @param dsigma The gradient with respect to standard deviation.
+   */
+  void ProbBackward(const arma::vec& observation, arma::vec& dmu,
+arma::vec& dsigma) const;
```
Comment (zoq) [inline src/mlpack/methods/ann/dists/normal_distribution.hpp:76]: We should put each parameter on a new line here.

File: src/mlpack/methods/ann/dists/normal_distribution.hpp (hunk 6 above)
Comment (nishantkr18) [reply src/mlpack/methods/ann/dists/normal_distribution.hpp:48]: Oh.. completely missed it

File: src/mlpack/methods/ann/dists/normal_distribution.hpp (hunk 4 above)
Comment (nishantkr18) [reply src/mlpack/methods/ann/dists/normal_distribution.hpp:29]: Yeah, sure

File: src/mlpack/tests/ann_dist_test.cpp [hunk 8]
```diff
@@ -127,4 +128,40 @@ BOOST_AUTO_TEST_CASE(JacobianBernoulliDistributionLogisticTest)
   }
 }
 
+/**
+ * Normal Distribution module test.
```
Comment (zoq) [inline src/mlpack/tests/ann_dist_test.cpp:132]: Should we run the Jacobian tests for the normal distribution as well?

Comment (nishantkr18) [reply src/mlpack/tests/ann_dist_test.cpp:133]: Oh yes.. definitely! I've made the changes.. pls have a look

**Comment by nishantkr18** [comment]
closed and reopened PR to re-trigger azure pipelines

File: src/mlpack/methods/ann/dists/normal_distribution.hpp [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * Implementation of the Normal Distribution function.
+ *
+ * Normal distribution is a function which accepts a mean and a standard deviation
+ * term and creates a probablity distribution out of it.
+ */
+template <typename DataType = arma::mat>
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/dists/normal_distribution.hpp:28]: Hey @nishantkr18, Could also please add a template parameter description. Thanks.

File: src/mlpack/methods/ann/dists/normal_distribution.hpp [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+   * Return a randomly generated observation according to the probability
+   * distribution defined by this object.
+   *
+   * @return Random observation from this Normal distribution.
+   */
+  DataType Sample() const;
+
+  /**
+   * Return the mean.
+   */
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/dists/normal_distribution.hpp:109]: Hey @nishantkr18, For single line comments, do you mind using `//` to comment. Thanks a lot!

File: src/mlpack/methods/ann/dists/normal_distribution.hpp [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+  const DataType& Mean() const { return mean; }
+
+  /**
+   * Return a modifiable copy of the mean.
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/dists/normal_distribution.hpp:113]:
```suggestion
   //! Modify the mean.
```

File: src/mlpack/methods/ann/dists/normal_distribution.hpp [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+   */
+  DataType& Mean() { return mean; }
+
+  /**
+   * Return the standard deviation.
+   */
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/dists/normal_distribution.hpp:119]:
```suggestion
   //! Get the standard deviation.
```

File: src/mlpack/methods/ann/dists/normal_distribution.hpp [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+  /**
+   * Serialize the distribution.
+   */
+  template<typename Archive>
+  void serialize(Archive& ar, const unsigned int /* version */)
+  {
+    // We just need to serialize each of the members.
+    ar & BOOST_SERIALIZATION_NVP(mean);
+    ar & BOOST_SERIALIZATION_NVP(sigma);
+  }
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/dists/normal_distribution.hpp:139]: I am sorry, it's a picky comment, Do you mind shifting this to the implementation file. Most of the code base follows that. Thanks a lot!

**Comment by nishantkr18** [comment]
>  I have left some minor style comments.

Thanks for those, I'll get them sorted. :+1: 

> Do you think that maybe we should update this PR and Bernoulli distribution in another PR and remove that section.

Yeah, that sounds like a plan. Again I'm waiting for @zoq or @rcurtin 's reviews. 
I think GaussianDistribution is used in a lot of places, and has API differences to NormalDistribution in `ann/dist`, similar is the case with Bernoulli dist. It'd be grt if we work on this in a separate PR since getting it merged might take some time. And I'd love to have a working normal-distribution just like this for now, so that we can get the #1912 PR moving. what do u say?

**Comment by kartikdutt18** [comment]
> And I'd love to have a working normal-distribution just like this for now, so that we can get the #1912 PR moving. what do u say?

Sure, I am fine with that. Maybe after this gets merged, we can open an issue to discuss the same, provided it needs discussing.

**Comment by nishantkr18** [comment]
> Maybe after this gets merged, we can open an issue to discuss the same, provided it needs discussing.

Yeah sure :+1:

File: src/mlpack/methods/ann/dists/normal_distribution.hpp [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+   */
+  DataType LogProbability(const DataType& observation) const;
+
+  /**
+   * Stores the gradient of the probabilities of the observations
+   * with respect to mean and standard deviation.
+   *
+   * @param observation The observation matrix.
+   * @param dmu The gradient with respect to mean.
+   * @param dsigma The gradient with respect to standard deviation.
```
Comment (zoq) [inline src/mlpack/methods/ann/dists/normal_distribution.hpp:72]:
```suggestion
   * @param dsigma The gradient with respect to the standard deviation.
```

File: src/mlpack/methods/ann/dists/normal_distribution.hpp [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+   */
+  void ProbBackward(const DataType& observation,
+                    DataType& dmu,
+                    DataType& dsigma) const;
+
+  /**
+   * Calculates the normal probability density function for each
+   * data point (column) in the given matrix.
+   *
+   * @param x List of observations.
```
Comment (zoq) [inline src/mlpack/methods/ann/dists/normal_distribution.hpp:82]: Do you think we should select a more descriptive parameter name here, like `observations`, we used `observation` before, so maybe it makes sense?

File: src/mlpack/methods/ann/dists/normal_distribution.hpp [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+  }
+
+  /**
+    * Calculates the log of normal probability density function for each
+    * data point (column) in the given matrix.
+    *
+    * @param x List of observations.
+    * @param log probabilities Output probabilities for each input observation.
+    */
+  void LogProbability(const DataType& x, DataType& probabilities) const
```
Comment (zoq) [inline src/mlpack/methods/ann/dists/normal_distribution.hpp:97]: See comment above.

File: src/mlpack/methods/ann/dists/normal_distribution.hpp [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+  {
+    probabilities = Probability(x);
+  }
+
+  /**
+    * Calculates the log of normal probability density function for each
+    * data point (column) in the given matrix.
+    *
+    * @param x List of observations.
+    * @param log probabilities Output probabilities for each input observation.
```
Comment (zoq) [inline src/mlpack/methods/ann/dists/normal_distribution.hpp:95]:
```suggestion
    * @param probabilities Output log probabilities for each input observation.
```

File: src/mlpack/methods/ann/dists/normal_distribution_impl.hpp [hunk 18]
```diff
@@ -0,0 +1,77 @@
+/**
+ * @file normal_distribution_impl.hpp
+ * @author xiaohong ji
+ * @author Nishant Kumar
+ *
+ * Implementation of the Normal distribution class.
```
Comment (zoq) [inline src/mlpack/methods/ann/dists/normal_distribution_impl.hpp:6]:
```suggestion
 * Implementation of the Normal Distribution class.
```

To be consistent, with what was used in the `normal_distribution.hpp` file.

File: src/mlpack/methods/ann/dists/normal_distribution_impl.hpp [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+  // Nothing to do here.
+}
+
+template<typename DataType>
+NormalDistribution<DataType>::NormalDistribution(
+    const DataType& mean,
+    const DataType& sigma) :
+    mean(mean),
+    sigma(sigma)
+{
```
Comment (zoq) [inline src/mlpack/methods/ann/dists/normal_distribution_impl.hpp:34]: Should we add a `// Nothing to do here.` here as well?

File: src/mlpack/methods/ann/dists/normal_distribution_impl.hpp [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+  return sigma * arma::randn<DataType>(mean.n_elem) + mean;
+}
+
+template<typename DataType>
+DataType NormalDistribution<DataType>::LogProbability(
+    const DataType& observation) const
+{
+  const DataType variance = arma::square(sigma);
+  DataType v1 = arma::log(sigma) + std::log(std::sqrt(2 * M_PI));
+  DataType v2 = arma::square(observation - mean) / (2 * variance);
```
Comment (zoq) [inline src/mlpack/methods/ann/dists/normal_distribution_impl.hpp:49]: I guess since `variance` is only used once we can inline the calculation here as well: `arma::square(observation - mean) / (2 * arma::square(sigma));`.

File: src/mlpack/methods/ann/dists/normal_distribution_impl.hpp [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+{
+  return sigma * arma::randn<DataType>(mean.n_elem) + mean;
+}
+
+template<typename DataType>
+DataType NormalDistribution<DataType>::LogProbability(
+    const DataType& observation) const
+{
+  const DataType variance = arma::square(sigma);
+  DataType v1 = arma::log(sigma) + std::log(std::sqrt(2 * M_PI));
```
Comment (zoq) [inline src/mlpack/methods/ann/dists/normal_distribution_impl.hpp:48]: Might be a good idea to use `const` for `v1, v2`.

File: src/mlpack/tests/ann_dist_test.cpp [hunk 22]
```diff
@@ -127,4 +129,174 @@ BOOST_AUTO_TEST_CASE(JacobianBernoulliDistributionLogisticTest)
...
+      jacobianB.col(k) = deltaMu % deriv;
+    }
+
+    BOOST_REQUIRE_LE(arma::max(arma::max(arma::abs(jacobianA - jacobianB))),
+        5e-3);
+  }
+}
+
+/**
+ * Jacobian Normal distribution module test for standard deviation.
```
Comment (zoq) [inline src/mlpack/tests/ann_dist_test.cpp:235]:
```suggestion
 * Jacobian Normal Distribution module test for standard deviation.
```

**Comment by nishantkr18** [comment]
> Don't have any comments, could you add to HISTORY.MD as well before we merge?

Done :+1:

**Comment by zoq** [comment]
Thanks again for the contribution!
